### ISSUE TITLE
complex numbers and include behavior

### ISSUE THREAD

**Issue by tonycurtis** [Sandia-OpenSHMEM/SOS#232]
Declaring a variable (any type) called "I" (capital-eye) prevents compilation:

```
#include <stdio.h>
#include <shmem.h>

int main()
{
    float I;

    shmem_init();
    shmem_finalize();

    return 0;
}
```

```
In file included from /home/tonyc/opt/crill/sandia/sos/include/shmem.h:20:0,
                 from test.c:2:
test.c: In function 'main':
test.c:7:11: error: expected identifier or '(' before '__extension__'
     float I;
           ^
```

**Comment by jdinan** [comment]
Thanks for the bug report!  This is because our shmem.h has ```#include <complex.h>``` to support the complex collectives.  I think we probably want to stop including that header and directly substitute ```_Complex``` for ```complex``` in the type names.

@jpdoyle what do you think?

**Comment by tonycurtis** [comment]
Or cf.

```
https://github.com/openshmem-org/openshmem/blob/master/src/shmem.h
```

lines 54-56 and 2230-2234.  Not sure which is the Right(tm) approach...

**Comment by jdinan** [comment]
One of the downsides of the spec's optimism that the same API is sufficient for C/C++ is that it's not clear whether we need two instances of each function for C++; one that takes e.g. ```float _Complex``` and another that takes ```std:complex<float>```.

**Comment by tonycurtis** [comment]
shmem.h turned into the "macrolympics"

**Comment by jpdoyle** [comment]
That's a thorny issue to deal with, but the solution in `https://github.com/openshmem-org/openshmem/blob/master/src/shmem.h` looks like a better workaround than using the raw types directly. Some of the C/C++ interaction around `complex.h` seems especially ill-defined and compiler-dependent.

**Comment by jdinan** [comment]
If the user tries to include ```complex.h``` after shmem.h undefines ```I```, will ```I``` get defined again?  Seems likely not, unless you also ```#undef __COMPLEX_H__``` which is nonstandard and likely to break.

**Comment by jpdoyle** [comment]
Including the complex collectives in `shmem.h` kinda prevents having a good solution to this. If you include `complex.h` without `#undef I` afterwards, you have the same problem we're seeing here. The C standard specifies that `I` is a macro for another macro, `_Imaginary_I`, so if someone wants to include `shmem.h`, define a variable named `I` *and* use `complex.h`'s imaginary constant, they can use `_Imaginary_I`.

**Comment by jdinan** [comment]
The C standard also says that ```complex``` must expand to ```_Complex```.  I was thinking that we can replace ```#include <complex.h>``` with:

```
#ifndef complex
#define complex _Complex
#define sos_undef_complex
#endif
...
#ifdef sos_undef_complex
#undef complex
#undef sos_undef_complex
#endif
```

**Comment by jpdoyle** [comment]
Then anything that uses `float complex` and `double complex` -- which are explicitly listed as types in the OpenSHMEM spec -- will break. Breaking the name for the `complex.h` `I` constant while leaving `T complex`/`std::complex<T>` intact seems less dangerous because `I` isn't used in the OpenSHMEM spec directly.

**Comment by jdinan** [comment]
Fair points -- I have another proposal.  We only declare the complex functions in ```shmem.h``` if the ```complex``` type is declared prior to inclusion of ```shmem.h```.  See #233 for implementation.

**Comment by jdinan** [comment]
Fixed in v1.4.0-dev
